### PR TITLE
Preserve texture bindings across framebuffer clears

### PR DESCRIPTION
**Pull request by bkaradzic-microsoft**
## Fix

`FrameBuffer::Clear()` calls bgfx `Encoder::touch()`, which discards encoder state, including texture bindings, both before and after its empty submission. Unlike that operation, a WebGL clear must preserve bound textures.

This breaks callers such as `ProceduralTexture.render()`: it binds its input textures, clears its destination, then draws with the inputs no longer bound.

Use the same empty-submission sequence with `BGFX_DISCARD_BINDINGS` and `BGFX_DISCARD_STATE` excluded from both discard masks. `setTexture()` records OpenGL sampler-unit indices as uniform writes, so preserving texture handles alone is insufficient: discarding the uniform writes makes both samplers read unit zero. Vertex/index/instance bindings and transform state are still discarded. This requires no texture-tracking cache, dependency changes, or Babylon.js changes.

## Coverage

Two GPU regressions verify two live sampler inputs across color-only, depth-only, empty-mask, and scissored combined clears, plus a real procedural texture mixing two inputs. They reuse the existing fullscreen-quad test helper.

Validated on upstream `d44847a8`, its pinned dependencies, and the lockfile's Babylon.js **9.21.2**, using Windows Release builds with D3D11 and GLES via ANGLE:

| Configuration | Result |
|---|---|
| D3D11, new regressions with unchanged renderer | Both fail |
| GLES, initial bindings-only fix | Both fail, reproducing Linux CI's `(32,64,96)` instead of `(80,112,144)` |
| D3D11, complete fix and related coverage | 24 pass |
| GLES, complete fix and related coverage | 19 pass, 2 existing backend-specific skips |

Related coverage includes the existing `ShaderCompilation.FragCoord*`, `ExternalTexture.*`, `Device.*`, `UniformPadding.*`, and `JavaScript.All` cases. The Linux failure exposed a missing part of the production fix; neither regression was weakened or skipped to address it.

Commands:

```text
cmake --build build\win32 --target UnitTests --config Release -- /m:2
UnitTests.exe --gtest_filter=NativeEngineClear.*:ShaderCompilation.FragCoord*:ExternalTexture.*:Device.*:UniformPadding.*:JavaScript.All
```

No visual references, tolerances, test exclusions, or dependency pins are changed.

### CODE REVIEW

File: Apps/UnitTests/Source/Tests.ShaderCompilation.FragCoord.cpp [hunk 1]
```diff
@@ -36,7 +36,8 @@ namespace
         uint32_t height,
         const std::string& vertexShader,
         const std::string& fragmentShader,
-        bool withInputTexture)
+        bool withInputTexture,
+        const std::string& setupScript = {})
```
[the remark this answers is not in the document]
Comment (bkaradzic-microsoft) [reply Apps/UnitTests/Source/Tests.ShaderCompilation.FragCoord.cpp:40]: `ShaderMaterial.setTexture()` adds a missing name to `this._options.samplers` before storing the texture; I verified that behavior in this branch's pinned Babylon.js 9.21.2. The injected setup runs before `scene.whenReadyAsync()` and effect compilation, so the constructor's initially empty sampler list is populated before binding.

Both regressions fail on unchanged upstream and pass with only the clear fix. The expected mixed RGBA also rejects either input being bound to both sampler units. An additional helper parameter is not needed here.

**Comment by bkaradzic-microsoft** [comment]
Fixed the Linux CI failure in 692f13493c98d0dc76d134be58925a84fa3d910b. bgfx stores OpenGL sampler-unit indices as uniform writes, so preserving texture bindings alone still discarded those indices. Both inputs then sampled unit zero.

I reproduced the exact Linux pixel failures locally with GLES via ANGLE, then preserved `BGFX_DISCARD_STATE` as well as `BGFX_DISCARD_BINDINGS`. The unchanged regressions and related coverage pass: D3D11 24/24; GLES 19 passing with 2 existing backend-specific skips. No assertions, tolerances, or exclusions were weakened. The updated CI matrix is running.

File: Apps/UnitTests/Source/Tests.ShaderCompilation.FragCoord.cpp [hunk 2]
```diff
@@ -258,6 +261,90 @@ namespace
...
+        "attribute vec3 position;\n"
+        "attribute vec2 uv;\n"
+        "varying vec2 vUV;\n"
+        "void main(void) { vUV = uv; gl_Position = vec4(position, 1.0); }\n";
+    const std::string fragmentShader =
+        "precision highp float;\n"
+        "uniform sampler2D inputSampler;\n"
+        "uniform sampler2D otherSampler;\n"
+        "varying vec2 vUV;\n"
+        "void main(void) { gl_FragColor = mix(texture2D(inputSampler, vUV), texture2D(otherSampler, vUV), 0.5); }\n";
```
[the remark this answers is not in the document]
Comment (bkaradzic-microsoft) [reply Apps/UnitTests/Source/Tests.ShaderCompilation.FragCoord.cpp:280]: I checked the actual pinned Babylon.js 9.21.2 implementation again: `ShaderMaterial.setTexture()` pushes a missing name into `this._options.samplers` before storing the texture. Both setup calls execute before `scene.whenReadyAsync()` and effect compilation, so both sampler names are registered in time.

This is also covered behaviorally: the tests reject both inputs sampling unit zero (the GLES bindings-only fix produced `(32,64,96)` instead of the expected `(80,112,144)`). The unchanged tests pass with the complete production fix on D3D11 and GLES. An additional sampler-list parameter is unnecessary.

The same registration concern was previously answered [here](https://github.com/BabylonJS/BabylonNative/pull/1869#discussion_r3984776661).

File: Core/Graphics/Source/FrameBuffer.cpp [hunk 3]
```diff
@@ -122,7 +122,10 @@ namespace Babylon::Graphics
         bgfx::setViewScissor(m_viewId.value());
         m_bgfxScissor = {};
 
-        encoder.touch(m_viewId.value());
+        // Keep texture bindings and uniform writes (including OpenGL sampler indices) across the empty draw.
+        constexpr uint8_t discardFlags{BGFX_DISCARD_ALL & ~(BGFX_DISCARD_BINDINGS | BGFX_DISCARD_STATE)};
```
[the remark this answers is not in the document]
Comment (bkaradzic-microsoft) [reply Core/Graphics/Source/FrameBuffer.cpp:126]: The pinned bgfx uses separate discard bits for these states: `BINDINGS=0x01`, `INDEX_BUFFER=0x02`, `INSTANCE_DATA=0x04`, `STATE=0x08`, `TRANSFORM=0x10`, and `VERTEX_STREAMS=0x20`. Our mask is `0xf6`, so it retains all four vertex/index/instance/transform discard bits.

`RenderBind::clear()` handles the sampler/resource binding slots; `RenderDraw::clear()` independently checks the four granular bits and clears their corresponding members. Thus retaining `BINDINGS` does not retain vertex/index/instance bindings in this dependency revision. No mask change is needed.

Source: [discard constants](https://github.com/BabylonJS/bgfx/blob/0837ead970a66875dcba37d1bd8e104db0879561/include/bgfx/defines.h#L257-L264), [separate binding and draw-state clearing](https://github.com/BabylonJS/bgfx/blob/0837ead970a66875dcba37d1bd8e104db0879561/src/bgfx_p.h#L2438-L2497).

File: Apps/UnitTests/Source/Tests.ShaderCompilation.FragCoord.cpp [hunk 4]
```diff
@@ -258,6 +261,90 @@ namespace
...
+    for (int clearMode = 0; clearMode < 4; ++clearMode)
+    {
+        SCOPED_TRACE(::testing::Message() << "clearMode=" << clearMode);
+        const std::string setupScript = R"(
+            var first = BABYLON.RawTexture.CreateRGBATexture(new Uint8Array([32, 64, 96, 255]), 1, 1, scene, false, false, 1);
+            var second = BABYLON.RawTexture.CreateRGBATexture(new Uint8Array([128, 160, 192, 255]), 1, 1, scene, false, false, 1);
+            material.setTexture("inputSampler", first);
+            material.setTexture("otherSampler", second);
```
[the remark this answers is not in the document]
Comment (bkaradzic-microsoft) [reply Apps/UnitTests/Source/Tests.ShaderCompilation.FragCoord.cpp:288]: Duplicate of the sampler-registration concern answered [here](https://github.com/BabylonJS/BabylonNative/pull/1869#discussion_r3984990583). In the pinned 9.21.2 implementation, these `setTexture()` calls add both names to the material's sampler list before effect compilation; the two-input pixel assertions also detect the unit-zero failure. No helper change is needed.

**Review by bghgary — APPROVED** [review]
[Reviewed by Copilot on behalf of @bghgary]

LGTM
